### ISSUE TITLE
Invalid segment entry size

### ISSUE THREAD

**Issue by lfam** [borgbackup/borg#1119]
I have been using Borg 1.0.3 to create archives on a remote append_only repo served by Borg 1.0.3.

Yesterday I began to get the following error for `create` and `list`. `check --repository-only` works. I haven't had time to try to restore anything yet.

Here is what I get while trying to `list` the repo. Unfortunately, I don't have the logs of the `create` where 
this might have started. I've replaced the name of the remote with 'remote'.

```
$ borg list --remote-path ~/.guix-profile/bin/borg remote:/srv/backup/leo/jasmine
Remote: Borg 1.0.3: exception in RPC call:
Remote: Traceback (most recent call last):
Remote:   File "/gnu/store/1h9xm0wpqv6m9hyns65hrcl4j0niqmr3-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/remote.py", line 96, in serve
Remote:     res = f(*args)
Remote:   File "/gnu/store/1h9xm0wpqv6m9hyns65hrcl4j0niqmr3-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/repository.py", line 445, in get
Remote:     return self.io.read(segment, offset, id_)
Remote:   File "/gnu/store/1h9xm0wpqv6m9hyns65hrcl4j0niqmr3-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/repository.py", line 665, in read
Remote:     size, tag, key, data = self._read(fd, self.put_header_fmt, header, segment, offset, (TAG_PUT, ))
Remote:   File "/gnu/store/1h9xm0wpqv6m9hyns65hrcl4j0niqmr3-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/repository.py", line 687, in _read
Remote:     segment, offset))
Remote: borg.helpers.IntegrityError: Invalid segment entry size [segment 9739, offset 234449]
Remote: Platform: Linux ficus 4.5.0-2-amd64 #1 SMP Debian 4.5.3-2 (2016-05-08) x86_64
Remote: Linux: debian stretch/sid 
Remote: Borg: 1.0.3  Python: CPython 3.4.3
Remote: PID: 14717  CWD: /home/leo
Remote: sys.argv: ['/gnu/store/1h9xm0wpqv6m9hyns65hrcl4j0niqmr3-borg-1.0.3/bin/.borg-real', 'serve', '--umask=077']
Remote: SSH_ORIGINAL_COMMAND: None
Remote: 
Data integrity error
Traceback (most recent call last):
  File "/gnu/store/yq8w362rdximwblza356m5aw5nhcwg1b-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/archiver.py", line 1520, in main
    exit_code = archiver.run(args)
  File "/gnu/store/yq8w362rdximwblza356m5aw5nhcwg1b-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/archiver.py", line 1465, in run
    return args.func(args)
  File "/gnu/store/yq8w362rdximwblza356m5aw5nhcwg1b-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/archiver.py", line 75, in wrapper
    kwargs['manifest'], kwargs['key'] = Manifest.load(repository)
  File "/gnu/store/yq8w362rdximwblza356m5aw5nhcwg1b-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/helpers.py", line 93, in load
    cdata = repository.get(cls.MANIFEST_ID)
  File "/gnu/store/yq8w362rdximwblza356m5aw5nhcwg1b-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/remote.py", line 354, in get
    for resp in self.get_many([id_]):
  File "/gnu/store/yq8w362rdximwblza356m5aw5nhcwg1b-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/remote.py", line 358, in get_many
    for resp in self.call_many('get', [(id_,) for id_ in ids], is_preloaded=is_preloaded):
  File "/gnu/store/yq8w362rdximwblza356m5aw5nhcwg1b-borg-1.0.3/lib/python3.4/site-packages/borgbackup-1.0.3-py3.4-linux-x86_64.egg/borg/remote.py", line 259, in call_many
    raise IntegrityError(res)
borg.helpers.IntegrityError: b'Remote Exception (see remote log for the traceback)'

Platform: Linux jasmine 4.5.0-2-amd64 #1 SMP Debian 4.5.3-2 (2016-05-08) x86_64 
Linux: debian stretch/sid 
Borg: 1.0.3  Python: CPython 3.4.3
PID: 18535  CWD: /home/leo
sys.argv: ['/gnu/store/yq8w362rdximwblza356m5aw5nhcwg1b-borg-1.0.3/bin/.borg-real', 'list', '--remote-path', '/home/leo/.guix-profile/bin/borg', 'remote:/srv/backup/leo/jasmine']
SSH_ORIGINAL_COMMAND: None
```

Any ideas? Can the repo be recovered? Even if not, I'm happy to help diagnose any bugs in Borg.

**Comment by enkore** [comment]
*Invalid segment entry size* means that the size field of an entry is way too small / way too large. This is likely caused by data corruption. Does SMART on the drives raise any flags?

**Comment by ThomasWaldmann** [comment]
http://paste.lisp.org/display/317347

**Comment by lfam** [comment]
I must have made a mistake when running `borg check` previously. It now reports the errors:

```
$ borg check -v remote:srv/backup/leo/jasmine --repository-only --remote-path ~/.guix-profile/bin/borg 
Remote: Starting repository check
Remote: Invalid segment entry size [segment 9608, offset 6659340]
Remote: Invalid segment entry size [segment 9609, offset 8493344]
Remote: Invalid segment entry size [segment 9610, offset 8760408]
Remote: Invalid segment entry size [segment 9611, offset 6874612]
Remote: Invalid segment entry size [segment 9612, offset 7720888]
Remote: Invalid segment entry size [segment 9613, offset 6088752]
Remote: Invalid segment entry size [segment 9614, offset 8744180]
Remote: Invalid segment entry size [segment 9615, offset 5615000]
Remote: Invalid segment entry size [segment 9616, offset 6049964]
Remote: Invalid segment entry size [segment 9617, offset 6319368]
Remote: Invalid segment entry size [segment 9618, offset 6417336]
Remote: Invalid segment entry size [segment 9619, offset 5739856]
Remote: Invalid segment entry size [segment 9620, offset 6620952]
Remote: Invalid segment entry size [segment 9621, offset 5506748]
Remote: Invalid segment entry size [segment 9685, offset 5269788]
Remote: Invalid segment entry size [segment 9686, offset 5250028]
Remote: Invalid segment entry size [segment 9687, offset 5243296]
Remote: Invalid segment entry size [segment 9688, offset 5249232]
Remote: Invalid segment entry size [segment 9689, offset 5255752]
Remote: Invalid segment entry size [segment 9690, offset 5244824]
Remote: Invalid segment entry size [segment 9719, offset 5641796]
Remote: Invalid segment entry size [segment 9720, offset 5252032]
Remote: Invalid segment entry size [segment 9721, offset 5243300]
Remote: Invalid segment entry size [segment 9722, offset 5247328]
Remote: Invalid segment entry size [segment 9723, offset 5261044]
Remote: Invalid segment entry size [segment 9724, offset 5243500]
Remote: Invalid segment entry size [segment 9725, offset 5689884]
Remote: Invalid segment entry size [segment 9726, offset 6524404]
Remote: Invalid segment entry size [segment 9727, offset 8795508]
Remote: Invalid segment entry size [segment 9728, offset 5287660]
Remote: Invalid segment entry size [segment 9729, offset 7214124]
Remote: Invalid segment entry size [segment 9730, offset 6872156]
Remote: Invalid segment entry size [segment 9731, offset 8172336]
Remote: Invalid segment entry size [segment 9732, offset 5495168]
Remote: Invalid segment entry size [segment 9733, offset 7415532]
Remote: Invalid segment entry size [segment 9734, offset 10152564]
Remote: Invalid segment entry size [segment 9735, offset 9634440]
Remote: Invalid segment entry size [segment 9736, offset 6927056]
Remote: Invalid segment entry size [segment 9737, offset 9165812]
Remote: Invalid segment entry size [segment 9738, offset 448062]
Remote: Invalid segment entry size [segment 9739, offset 5732565]
Remote: Index object count mismatch. 979295 != 961597
Remote: Completed repository check, errors found.
```

**Comment by lfam** [comment]
Regarding SMART:

```
# smartctl -a /dev/sda1                                                                                                                                                             
smartctl 6.3 2014-07-26 r3976 [x86_64-linux-4.5.0-2-amd64] (local build)
Copyright (C) 2002-14, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Vendor:               ASMT
Product:              2115
Revision:             0
Compliance:           SPC-4
User Capacity:        500,107,862,016 bytes [500 GB]
Logical block size:   512 bytes
Logical Unit id:      0x5000000000000001
Serial number:        E2D987654321
Device type:          disk
Local Time is:        Thu Jun  2 20:09:02 2016 EDT
SMART support is:     Available - device has SMART capability.
SMART support is:     Enabled
Temperature Warning:  Disabled or Not Supported

=== START OF READ SMART DATA SECTION ===
SMART Health Status: OK

Error Counter logging not supported

Device does not support Self Test logging
```

**Comment by lfam** [comment]
@enkore N=9756

**Comment by lfam** [comment]
I realized that I gave smartctl a partition rather than the full device as an argument, so there was not much to look at. On #borgbackup TW and I looked at the full report. No egregious problems seen.

`borg check --repair` made it possible to use the repo, so I looked at the archive where the problems started. There are corrupt files in that archive, for sure.

I'll close this bug for now.

**Comment by ThomasWaldmann** [comment]
Well, while there were no pending or reallocated sectors, there were some UDMA CRC / ICRC errors and also the acceleration / free-fall sensor was triggered some times, so it is the question whether some hardware issue impacted the backup.

If not, it could be also the case that some other issue brought the repo into that bad state - the sequence of events for that would be interesting, if it could be determined.

**Comment by lfam** [comment]
If anybody has any ideas about how to reproduce this sort of failure, I'm happy to try them out.

**Comment by enkore** [comment]
Stapling #1138 for reference.

While it wasn't established that the errors were actually logged and therefore mis-reported, it was possibly the case given the SMART errors observed.

**Comment by taladar** [comment]
I have the "Invalid segment entry size" on a new regular (not append-only) repository created with Borg 1.0.6. I deleted the repo and started the backup again in a new repo and the same error occurred (though as far as I recall not exactly on the same segment and offset). Dozens of other computers are backed up to the same disk using both attic and Borg without problems. There are no SMART errors or other disk errors in dmesg on the source or target host.

`Remote: Borg 1.0.6: exception in RPC call:
Remote: Traceback (most recent call last):
Remote:   File "borg/remote.py", line 97, in serve
Remote:   File "borg/repository.py", line 458, in get
Remote:   File "borg/repository.py", line 160, in get_transaction_id
Remote:   File "borg/repository.py", line 322, in replay_segments
Remote:   File "borg/repository.py", line 331, in _update_index
Remote:   File "borg/repository.py", line 648, in iter_objects
Remote:   File "borg/repository.py", line 704, in _read
Remote: borg.helpers.IntegrityError: Invalid segment entry size [segment 8976, offset 2307863]
Remote: Platform: Linux targethost 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt25-2+deb8u3 (2016-07-02) x86_64
Remote: Linux: debian 8.5 
Remote: Borg: 1.0.6  Python: CPython 3.5.2
Remote: PID: 30604  CWD: /home/attic_hostname
Remote: sys.argv: ['borg', 'serve', '--umask=077']
Remote: SSH_ORIGINAL_COMMAND: None
Remote: 
Data integrity error
Traceback (most recent call last):
  File "borg/archiver.py", line 1609, in main
  File "borg/archiver.py", line 1546, in run
  File "borg/archiver.py", line 75, in wrapper
  File "borg/helpers.py", line 106, in load
  File "borg/remote.py", line 367, in get
  File "borg/remote.py", line 371, in get_many
  File "borg/remote.py", line 278, in call_many
  File "borg/remote.py", line 259, in handle_error
borg.helpers.IntegrityError: b'Remote Exception (see remote log for the traceback)'

Platform: Linux hostname 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt25-2+deb8u3 (2016-07-02) x86_64 
Linux: debian 8.5 
Borg: 1.0.6  Python: CPython 3.5.2
PID: 32545  CWD: /root
sys.argv: ['/etc/scripts/borg', 'create', '-s', '--exclude', '/proc', '--exclude', '/dev', '--exclude', '/sys', '--exclude', '/backup/[!_]*', '--exclude', '/chroot', '--exclude', '/mnt', '--exclude', '/var/cache/
apt', '--exclude', '/var/lib/apt/lists', '--exclude', '/usr/local/tarsnap-cache', '--exclude', '/var/cache/apt-cacher-ng', 'attic_hostname@targethost.ourdomain.com:/backup/attic_store_location::hostname.ourdomain.com_20
160725_150305', '/']
SSH_ORIGINAL_COMMAND: None`

I replaced the hostnames and domains but kept the log otherwise unchanged.

**Comment by ThomasWaldmann** [comment]
Was there some other borg operation running against the same repository?

**Comment by taladar** [comment]
Since I am calling borg from a script that does its own locking and I am using a separate repository per host I do not think there were any other borg processes running in parallel.

**Comment by enkore** [comment]
You may have to consider the possibility that this is real data corruption on the disks/controllers...

**Comment by taladar** [comment]
I find it highly unlikely that the disk would fail in the way to exactly reproduce the issue in two consecutive borg backup runs just for this client system (with full deletion of the repo in between) but not on any of the 50+ other machines backing up to the same server. Particularly with no SMART errors or dmesg lines indicating any errors.

**Comment by ThomasWaldmann** [comment]
Related code:
```
        if size > MAX_OBJECT_SIZE or size < fmt.size:
            raise IntegrityError('Invalid segment entry size [segment {}, offset {}]'.format(
                segment, offset))
```
So, the `size` value in the header was outside of the valid range.
Likely it was either 0 (too small) or some more-or-less random bitstring (resulting in an integer bigger than 20MiB).

**Comment by enkore** [comment]
*scratches head*

I recently calculated how many archives it would need to fill the manifest beyond 20 MiB, irrc ~300k archives. So that's probably not it.

**Comment by enkore** [comment]
> The highlevel operation that failed was the loading of the manifest - this is the first (read) access to the repository.

Yeah, but it hits the object during segment replay:

```
Remote: File "borg/remote.py", line 97, in serve
Remote: File "borg/repository.py", line 458, in get
Remote: File "borg/repository.py", line 160, in get_transaction_id
Remote: File "borg/repository.py", line 322, in replay_segments
Remote: File "borg/repository.py", line 331, in _update_index
Remote: File "borg/repository.py", line 648, in iter_objects
Remote: File "borg/repository.py", line 704, in _read
```

So the commit state wasn't clean at that point, and it doesn't need to be the manifest object causing the error.

**Comment by ThomasWaldmann** [comment]
Somehow easiest explanation would be if due to some malfunction either some additional crap was written into a segment file or some valid write data was not written to a segment file - both causing data that is not a entry header at a place where a header is expected.

**Comment by enkore** [comment]
I would still not rule IO errors out. @taladar could try e.g. deleting the repo again and copying one of the other repos over it. If the allocators recycles the block like it probably did when the repo was first removed and then recreated you may see the error there again. This isn't a sure test, though.

**Comment by ThomasWaldmann** [comment]
@taladar could you try:
- run the borg init / borg create again, log the full commands you used and also use -v --show-rc for logging at INFO level, keep all stdout / stderr output
- maybe try reproducing with a smaller (non-confidential) set of files, like /bin and/or /usr maybe.
- backing up the problematic client locally - do you run into same issues? if you can reproduce the issue locally, analyzing/debugging that is way easier.

**Comment by taladar** [comment]
I did run into the same issue running the same backup locally. The error only occurs after quite a long time and judging by the progress indicator still visible in no particular spot either (somewhere in our /var/lib/docker btrfs filesystem which is the majority of the data on the server right now).

I do not see how to minimize this easily due to the long runtime and the rather arbitrary spot where the error seems to occur (the backup via SSH failed at about 42GB of data in the repository).

```
~ # ionice -c 3 /etc/scripts/borg create -s -v -p --show-rc --exclude /proc --exclude /dev --exclude /sys --exclude '/backup/[!_]*' --exclude /chroot --exclude /mnt --exclude /run --exclude /var/cache/apt --exclude /var/lib/apt/lists --exclude /usr/local/tarsnap-cache --exclude /var/cache/apt-cacher-ng /backup/borg-backup::tkrenn2.saltation.de_$(date '+%Y%m%d_%H%M%S') /                                                    
Data integrity error49.22 GB D 31424092 N var/lib/docker/btrfs/subvolumes/8794c1fa3af8f1528b11e02d71656ee8b6ba89d13f071262109558d03c1585ea/usr/lib/i386-linux-gnu/libORBit-2.so.0.1.0                               
Traceback (most recent call last):
  File "borg/archiver.py", line 1609, in main
  File "borg/archiver.py", line 1546, in run
  File "borg/archiver.py", line 81, in wrapper
  File "borg/archiver.py", line 247, in do_create
  File "borg/archiver.py", line 219, in create_inner
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 301, in _process
  File "borg/archiver.py", line 275, in _process
  File "borg/archive.py", line 689, in process_file
  File "borg/archive.py", line 281, in add_item
  File "borg/archive.py", line 285, in write_checkpoint
  File "borg/archive.py", line 317, in save
  File "borg/repository.py", line 199, in commit
  File "borg/repository.py", line 287, in compact_segments
  File "borg/repository.py", line 648, in iter_objects
  File "borg/repository.py", line 704, in _read
borg.helpers.IntegrityError: Invalid segment entry size [segment 8703, offset 1412154]

Platform: Linux tkrenn2 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt25-2+deb8u3 (2016-07-02) x86_64 
Linux: debian 8.5 
Borg: 1.0.6  Python: CPython 3.5.2
PID: 10019  CWD: /root
sys.argv: ['/etc/scripts/borg', 'create', '-s', '-v', '-p', '--show-rc', '--exclude', '/proc', '--exclude', '/dev', '--exclude', '/sys', '--exclude', '/backup/[!_]*', '--exclude', '/chroot', '--exclude', '/mnt', '--exclude', '/run', '--exclude', '/var/cache/apt', '--exclude', '/var/lib/apt/lists', '--exclude', '/usr/local/tarsnap-cache', '--exclude', '/var/cache/apt-cacher-ng', '/backup/borg-backup::tkrenn2.saltation.de_20160810_105544', '/']
SSH_ORIGINAL_COMMAND: None

terminating with error status, rc 2
zsh: exit 2     ionice -c 3 /etc/scripts/borg create -s -v -p --show-rc --exclude /proc  /dev
ionice -c 3 /etc/scripts/borg create -s -v -p --show-rc --exclude /proc  /dev  36592.16s user 4739.64s system 71% cpu 16:01:20.92 total
```

I am sorry I took this long to reply but as you can see the backup run takes quite a while.

**Comment by enkore** [comment]
> 31424092 N

This looks like #1452

**Comment by ThomasWaldmann** [comment]
Duplicate of #1452.
